internal/pkg/sourceinfer/analyzer.go

```diff
@@ -0,0 +1,336 @@
+	"golang.org/x/tools/go/ssa"
+)
+
+// ResultType is a set of types.Object that are inferred Sources.
+type ResultType map[types.Object]bool
+
+type inferredSource struct{}
+
+func (i inferredSource) AFact() {}
+
+func (i inferredSource) String() string {
+	return "inferred source"
+}
+
+var Analyzer = &analysis.Analyzer{
+	Name: "sourceinfer",
+	Doc: `This analyzer infers source types from typedefs and fields.
+
+Suppose Foo has been configured to be a source type.
+
+Then, the following type definition would cause Bar to be identified
+as a source, because its underlying type is Foo:
+type Bar Foo
+Indeed, if Foo contains sensitive data, then so does Bar, because they have
+the same fields.
+
+Similarly, the following type definition would cause Baz to be identified
+as a source, because it holds a field of type Foo:
+type Baz struct {
+  f Foo
+}
+Indeed, if Foo contains sensitive data, and Bar contains Foo, Bar also
+contains that data, via Foo.
+
+The analysis is recursive: types identified as sources during analysis
+will be used to evaluate the sourceyness of types that depend on them.
+For example, in the below type definitions, both Qux and Quux will be
+identified as sources:
+type Qux Foo
+type Quux Qux
+`,
+	Run: run,
+	Requires: []*analysis.Analyzer{
+		inspect.Analyzer,
+		buildssa.Analyzer,
+	},
+	ResultType: reflect.TypeOf(new(ResultType)).Elem(),
+	FactTypes:  []analysis.Fact{new(inferredSource)},
+}
+
+type ObjectGraph map[types.Object][]types.Object
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	conf, err := config.ReadConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	ins := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	typeGraph := createTypeDefGraph(pass, ins)
+
+	builtSSA := pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA)
+	addFieldEdges(builtSSA, typeGraph)
+
+	inferredSources := inferSources(pass, conf, typeGraph)
+
+	return inferredSources, nil
+}
+
+func createTypeDefGraph(pass *analysis.Pass, ins *inspector.Inspector) ObjectGraph {
+	defGraph := ObjectGraph{}
+
+	genDeclFilter := []ast.Node{
+		(*ast.GenDecl)(nil),
+	}
+
+	ins.Preorder(genDeclFilter, func(n ast.Node) {
+		gd, ok := n.(*ast.GenDecl)
+		if !ok {
+			return
+		}
+		for _, spec := range gd.Specs {
+			// type definition
+			ts, ok := spec.(*ast.TypeSpec)
+			if !ok {
+				continue
+			}
+
+			// type alias, e.g. type A = B
+			// we don't need to handle these, since the type will be inlined
+			if ts.Assign != token.NoPos {
+				continue
+			}
+
+			// struct definition, fields will be checked later in the analysis
+			if _, ok := ts.Type.(*ast.StructType); ok {
+				continue
+			}
+
+			typeBeingDefined := pass.TypesInfo.ObjectOf(ts.Name)
+
+			selectorFinder := &selectorFinder{nil, map[*ast.Ident]bool{}}
+			ast.Walk(selectorFinder, (ast.Node)(ts.Type))
+			for _, sel := range selectorFinder.foundSelectors {
+				named, ok := pass.TypesInfo.TypeOf(sel).(*types.Named)
+				if !ok {
+					continue
+				}
```

Comment:
You are correct. The current approach is restricted to named types. For a case like this, I think instead of returning a slice of `types.Object`, we may want to return a `IsSource` function, which in this case would scan the fields for named types and determine if any of them was identified as a source during the analysis.

To me, it would make more sense to add that functionality in a future PR, since this PR is already quite big .